crates/types/src/consensus.rs

```diff
@@ -372,28 +373,32 @@ impl<PAYLOAD: BlockPayload> BlockPayloadStore<PAYLOAD> {
     /// multiple calls to [`insert`](Self::insert) for the same payload commitment result in
     /// multiple owning references to the payload commitment. [`remove`](Self::remove) must be
     /// called once for each reference before the payload commitment will be deallocated.
-    pub fn insert(&mut self, payload: PAYLOAD) {
+    ///
+    /// # Errors
+    /// If the transaction length conversion fails.
+    pub fn insert(&mut self, payload: PAYLOAD) -> Result<(), PAYLOAD::Error> {
         self.0
-            .entry(payload.commit())
+            .entry(vid_commitment(payload.clone().encode()?.collect()))
```

Comment:
I don't think we should compute the VID commitment here. This is called from the DA task where we already have it. So just add it as a parameter to `insert`

Reply:
In fact, maybe we shouldn't store `PAYLOAD` here at all. Maybe it should just be `VidCommitment -> Vec<u8>`. Because `PAYLOAD` is derived from the `Vec<u8>` _plus_ the metadata. But the DA task isn't really promising anything about the metadata (I don't think it signs the metadata when it votes, does it?) The DA task is really promising "I will make available a certain sequence of bytes with this VID commitment". So maybe this whole mapping should deal in bytes, and we reconstruct the `PAYLOAD` from the bytes and the metadata (which comes from the block header) when we emit the decide event, or possibly when the query service consumes the decide event.

WDYT @shenkeyao ?

Reply:
I think maybe if we did that, it would clean up error handling a bit too, since we don't have to constantly go back and forth between bytes and `PAYLOAD` -- everything is bytes, right up until the very last step.

Reply:
> I don't think we should compute the VID commitment here. This is called from the DA task where we already have it.

Just wanted to note that besides the DA task, we also call this function with the payload instead of the VID commitment in `crates/hotshot/src/lib.rs`. But I agree we change the input from the payload to VID commitment, and in places where we have the payload, we can just compute its commitment.

> Maybe it should just be VidCommitment -> Vec<u8>.

Agreed!

Reply:
I have trouble following everything. Is this payload store maintained only by DA committee nodes, or everyone? In any case, Jeb's suggestion to store only the bytes instead of the payload struct makes sense to me.

Perhaps it's confusing because you cannot get the bytes without first creating a payload. Yet it sounds like all we really want is the bytes. So the flow becomes:
1. build payload from txs
2. encode to bytes, forget about the payload, use bytes everywhere until the end.
3. re-construct the payload at the end.

If that's the case then perhaps we should rethink the payload API?

Finally: as a general rule to everyone: please do not clone payloads, do not call `encode()?.collect()` more than once, do not compute the commitment more than once. If you find the need to violate this constraint then there's something wrong with our design and we need to fix it.

Reply:
This store is maintained by everyone but I think after genesis it's only populated when you're on the DA committee. It's only temporary, so that nodes who are on DA can include the entire payload in the `Decide` events, whereafter they become the responsibility of the consumer (e.g. query service) to persist. So entries are created when a DA member signs a block and deleted after that block is finalized, or after it gets forked out of the chain.

Regarding the workflow, it is true that on the builder side, we construct the block, encode it, and then immediately forget the structured representation. On the other end, we do want the structured payload: HotShot will construct it right before emitting a decide event, use the `Blockpayload` interface to update its mempool statistics, and then send it off to the query service, where we definitely want it structured.

Maybe we can come up with a better API for the builder, like `encode_txns(Vec<Transaction>) -> Vec<u8>`, skipping the intermediate step, or `into_bytes(self) -> Vec<u8>` (keeping the intermediate step, but moving out of `self` to ensure there is no unnecessary cloning). But I don't think this is a high priority for right now, since there is actually not very much cost to `encode()` once you have a `BlockPayload` (and once we fix the API to return `Cow` or have Jellyfish take an iterator)